gap/PolygonalComplexes/properties.gi

```diff
@@ -218,6 +218,141 @@ InstallMethod( FaceCounter, "for a twisted polygonal complex",
         return Collected( Compacted( faceDegrees ) );
     end
 );
+
+InstallMethod( ButterflyCounter, "for a simplicial surface",
+    [IsSimplicialSurface],
+    function(surface)
+	local e,temp2,counter,g,VerticesOfOrthogonalEdge,voe1,voe2,degOfVert1,degOfVert2;
+
+        VerticesOfOrthogonalEdge:=function(S,e)
+            local temp;
+            temp:=FacesOfEdge(S,e);
+            temp:=Union(VerticesOfFace(S,temp[1]),VerticesOfFace(S,temp[2]));
+            return Difference(temp, VerticesOfEdge(S,e));
+	end;
+	counter:=[];
+	for e in InnerEdges(surface) do 
+		voe1:=VerticesOfEdge(surface,e);
+		degOfVert1:=SortedList(List(voe1,g->FaceDegreeOfVertex(surface,g)));
+		voe2:=VerticesOfOrthogonalEdge(surface,e);
+		degOfVert2:=SortedList(List(voe2,g->FaceDegreeOfVertex(surface,g)));
+		Add(counter,[degOfVert1,degOfVert2]);
+	od;
+	return Set(Collected(counter));
+end
+);
+
+
+InstallMethod( UmbrellaCounter, "for a simplicial surface",
+    [IsSimplicialSurface],
+    function(surface)
+	local n,counter,v,temp,orb,tup,G,perm,OrbitOnList,tempcounter,temp1,rightumb,bub;
+	tempcounter:=[];
+	if not IsClosedSurface(surface) then 
+		return fail;
+	fi;
+	rightumb:=function(L)
+		local m,temp,i,t;
+		temp:=orb;		
+		n:=Length(temp[1]);
+		i:=2;
+		m:=Minimum(temp[1]);
+		temp:=Filtered(temp,g->g[1]=m);
+		while temp <>[] and i<=n do 
+			m:=Minimum(List(temp,g->g[i]));
+			temp:=Filtered(temp,g->g[i]=m);
+			i:=i+1;
+			if Length(temp)=1 then
+				return temp[1];
+			fi; 	
+		od;
+		return temp[1];
+	end;
+
+	OrbitOnList:=function(G,L)
+		local g,orb,temp,temp1,i,l,n;
+		orb:=[];
+		n:=Length(L);
+		l:=[1..n];
+		temp1:=[];
+		for g in G do
+			temp:=[]; 
+			for i in [1..n] do
+				temp[i]:=L[i^g];
+			od;
+			Add(temp1,temp);
+		od;
+		return temp1;
+	end;
+	bub:=function(list)
+		local g,tempm,tempL,m;
+		tempL:=[];
+		tempm:=List(list,g->Length(g));
+		for m in Set(tempm) do
+			temp:=Filtered(list,g->Length(g)=m);
+			Append(tempL,temp);
+		od; 
+		return tempL;
+	end;
+	for v in Vertices(surface) do
+		temp:=UmbrellaPathOfVertex(surface,v);
+		temp:=EdgesAsList(temp);
+		temp:=List([1..Length(temp)-1],g->temp[g]);
+		temp:=List(temp,g->Difference(VerticesOfEdge(surface,g),[v])[1]);
+		temp:=List(temp,g->FaceDegreeOfVertex(surface,g));
+		Add(tempcounter,temp);
+	od;
+	counter:=[];
+	
+	while tempcounter <> [] do 
```

Comment:
From here on it's hard for me to understand what the program is doing. Some comments would be helpful in particular for the methods: rightumb, OrbitOnLists, bub.